File: py/src/axon/schema.py
"""Draft-07 JSON Schema validation for spec/ documents.

All 47 schema files in spec/ declare
``"$schema": "http://json-schema.org/draft-07/schema#"`` (contract notes
S5). ``jsonschema.Draft7Validator`` implements draft-07 semantics, confirmed
against this exact schema shape: ``if``/``then`` (host.schema.json's one
conditional), ``$ref`` + ``definitions`` (events/capabilities/invoke
schemas), ``enum``, ``pattern``, ``uniqueItems``, ``minItems``, and
``additionalProperties`` all validate as draft-07 specifies with no extra
registration needed.
"""

from __future__ import annotations

import json
from typing import Any

from jsonschema import Draft7Validator

from axon._spec import read_spec_file
from axon.errors import SchemaError

_compiled_cache: dict[str, Draft7Validator] = {}


def _compile(schema_path: str) -> Draft7Validator:
    cached = _compiled_cache.get(schema_path)
    if cached is not None:
        return cached

    try:
        schema = json.loads(read_spec_file(schema_path))
    except (OSError, json.JSONDecodeError) as err:
        raise SchemaError(f"read/parse schema {schema_path}: {err}") from err

    try:
        Draft7Validator.check_schema(schema)
        validator = Draft7Validator(schema)
    except Exception as err:  # noqa: BLE001 - surfaced as our own sentinel
        raise SchemaError(f"compile schema {schema_path}: {err}") from err

    _compiled_cache[schema_path] = validator
    return validator


def validate(path: str, schema_path: str, doc: Any) -> None:
    """Validates doc (already parsed from YAML or JSON) against schema_path
    (relative to the spec root) as draft-07. Raises SchemaError on any
    violation; path is carried in the error message for diagnostics only.
    """
    validator = _compile(schema_path)
    errors = sorted(validator.iter_errors(doc), key=lambda e: list(e.path))
    if errors:
        details = "; ".join(f"{'/'.join(str(p) for p in e.path) or '/'} {e.message}" for e in errors)
        raise SchemaError(f"{path} violates {schema_path}: {details}")

File: ts/src/schema.ts
import Ajv, { type Schema, type ValidateFunction } from "ajv";
import { readSpecFile } from "./spec";
import { SchemaError } from "./errors";

// All 47 schema files in spec/ declare
// "$schema": "http://json-schema.org/draft-07/schema#" (contract notes
// §5). ajv v8's default `Ajv` class (as opposed to its `Ajv2019`/`Ajv2020`
// subclasses) targets draft-07 semantics, confirmed empirically against
// this exact schema shape: `if`/`then` (host.schema.json's one
// conditional), `$ref` + `definitions` (events/capabilities/invoke
// schemas), `enum`, `pattern`, `uniqueItems`, `minItems`, and
// `additionalProperties` all validate as draft-07 specifies with no
// extra metaschema registration needed.
const ajv = new Ajv({ strict: false });

const compiledCache = new Map<string, ValidateFunction>();

function compile(schemaPath: string): ValidateFunction {
  const cached = compiledCache.get(schemaPath);
  if (cached) return cached;

  let schema: Schema;
  try {
    schema = JSON.parse(readSpecFile(schemaPath));
  } catch (err) {
    throw new SchemaError(`read/parse schema ${schemaPath}: ${(err as Error).message}`);
  }

  let compiled: ValidateFunction;
  try {
    compiled = ajv.compile(schema);
  } catch (err) {
    throw new SchemaError(`compile schema ${schemaPath}: ${(err as Error).message}`);
  }
  compiledCache.set(schemaPath, compiled);
  return compiled;
}

/**
 * Validates doc (already parsed from YAML or JSON) against schemaPath
 * (relative to the spec root) as draft-07. Throws SchemaError on any
 * violation; path is carried in the error message for diagnostics only.
 */
export function validate(path: string, schemaPath: string, doc: unknown): void {
  const validateFn = compile(schemaPath);
  if (!validateFn(doc)) {
    const details = validateFn.errors?.map((e) => `${e.instancePath || "/"} ${e.message}`).join("; ");
    throw new SchemaError(`${path} violates ${schemaPath}: ${details}`);
  }
}
